### PR TITLE
gh-111495: Add tests for PyTuple C API

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-111495

### CODE REVIEW

**Comment by skirpichev** [comment]
Probably, could skip news:)

It's https://github.com/python/cpython/pull/111606, with some minor fixes and tests for PyTuple_Pack, PyTuple_SetItem/PyTuple_SET_ITEM and PyTuple_Resize.  Limited API C functions are moved to _testlimitedcapi/ directory.

@serhiy-storchaka

**Comment by skirpichev** [comment]
@rawwar, let me know if you are not OK with this using your code.

File: Lib/test/test_capi/test_tuple.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        self.assertEqual(get_item(tup, 0), 1)
+        self.assertEqual(get_item(tup, 2), 3)
+
+    def test_tuple_getslice(self):
+        # Test PyTuple_GetSlice()
+        getslice = _testlimitedcapi.tuple_getslice
+
+        tup = (1, 2, 3)
+
+        # empty
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:142]: Test also with start == stop and start > stop (if it is not an undefined behavior).

File: Modules/_testlimitedcapi/tuple.c [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+static PyObject *
+tuple_new(PyObject* Py_UNUSED(module), PyObject *len)
+{
+    return PyTuple_New(PyLong_AsSsize_t(len));
+}
+
+static PyObject *
+tuple_pack(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    Py_ssize_t size = PyLong_AsSize_t(PyTuple_GetItem(args, 0));
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c]: Please add error checks for all C API.

I think that it is simpler to use `PyArg_ParseTuple(args, "n|OO", &arg1, &arg2)`.

File: Lib/test/test_capi/test_tuple.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+        tup = (1, 2, 3)
+        self.assertEqual(getitem(tup, 0), 1)
+        self.assertEqual(getitem(tup, 2), 3)
+
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MIN)
+        self.assertRaises(IndexError, getitem, tup, -1)
+        self.assertRaises(IndexError, getitem, tup, len(tup))
+        self.assertRaises(IndexError, getitem, tup, PY_SSIZE_T_MAX)
+        self.assertRaises(SystemError, getitem, 42, 1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:120]: Test also with a list.
```suggestion
        self.assertRaises(SystemError, getitem, [1, 2, 3], 1)
        self.assertRaises(SystemError, getitem, 42, 1)
```

File: Lib/test/test_capi/test_tuple.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+        # slice
+        self.assertEqual(getslice(tup, 1, 3), (2, 3))
+
+        # whole
+        self.assertEqual(getslice(tup, 0, 3), tup)
+        self.assertEqual(getslice(tup, 0, 100), tup)
+        self.assertEqual(getslice(tup, -100, 100), tup)
+
+        # CRASHES getslice(NULL, 0, 0)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:175]: Test also with a list and with a non-sequence.

It would be nice to test also with a TupleSubclass, but this is more difficult.

File: Lib/test/test_capi/test_tuple.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        tup = (0, 0)
+        self.assertEqual(setitem(tup, 0, 1), (1, 0))
+        self.assertEqual(setitem(tup, 1, 1), (0, 1))
+
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MIN, 1)
+        self.assertRaises(IndexError, setitem, tup, -1, 1)
+        self.assertRaises(IndexError, setitem, tup, len(tup), 1)
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MAX, 1)
+
+        # CRASHES setitem(NULL, 1, 5)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]: Test also with a list and with a non-sequence.

**Comment by rawwar** [comment]
> @rawwar, let me know if you are not OK with this using your code.

Please reuse it. I am completely fine with it.

File: Lib/test/test_capi/test_tuple.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    def test_tuple_set_item(self):
+        # Test PyTuple_SET_ITEM()
+        set_item = _testcapi.tuple_set_item
+
+        tup = (0, 0)
+        self.assertEqual(set_item(tup, 0, 1), (1, 0))
+        self.assertEqual(set_item(tup, 1, 1), (0, 1))
+
+    def test_tuple_resize(self):
+        # Test PyTuple_Resize()
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]: A more complex test is needed here. `PyTuple_Resize()` should be tested with a non-empty tuple (and also with non-tuples). It can also be used to shrink the tuple. There are also two different cases -- refcount == 1, and refcount > 1, so you need either two wrappers, or a boolean parameter which specify whether to make a fresh copy of the input tuple or use the input argument itself. See how it was done in other tests.

File: Modules/_testlimitedcapi/tuple.c (hunk 2 above)
Comment (skirpichev) [reply Modules/_testlimitedcapi/tuple.c]: > Please add error checks for all C API.

I hope I did that.  Missing case is something like ``pack(2, 1)``, which crashes now.

File: Lib/test/test_capi/test_tuple.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        tup = (0, 0)
+        self.assertEqual(setitem(tup, 0, 1), (1, 0))
+        self.assertEqual(setitem(tup, 1, 1), (0, 1))
+
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MIN, 1)
+        self.assertRaises(IndexError, setitem, tup, -1, 1)
+        self.assertRaises(IndexError, setitem, tup, len(tup), 1)
+        self.assertRaises(IndexError, setitem, tup, PY_SSIZE_T_MAX, 1)
+        self.assertRaises(SystemError, setitem, [1], 0, 1)
+        self.assertRaises(SystemError, setitem, 42, 0, 1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]: These errors are raised not by the tested function, but by the improper use of the C API in the wrapper. You should only make a new tuple if the argument is a tuple.

File: Modules/_testlimitedcapi/tuple.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+tuple_setitem(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    PyObject *obj, *value, *newtuple;
+    Py_ssize_t i;
+    if (!PyArg_ParseTuple(args, "OnO", &obj, &i, &value)) {
+        return NULL;
+    }
+    NULLABLE(obj);
+    NULLABLE(value);
+    if (obj) {
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c]:
```suggestion
    if (obj & PyTuple_CheckExact(obj)) {
```
or
```suggestion
    if (new) {
```
where `new` is an optional boolean parameter (see the test for `PyBytes_Concat` for example). This would allow to test refcount == 1 and refcount > 1.

File: Lib/test/test_capi/test_tuple.py (hunk 6 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_tuple.py]: I hope, I did that.  Unfortunately, it seems not too much could be done for new=False scenario.  Here we could test refcount > 1 condition.

File: Lib/test/test_capi/test_tuple.py (hunk 7 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_tuple.py]: Yep, fixed (with a slight modification of your proposal).

File: Modules/_testlimitedcapi/tuple.c [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SetItem(newtuple, n, Py_XNewRef(PyTuple_GetItem(obj, n)));
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
+    }
+    if (PyTuple_SetItem(newtuple, i, Py_XNewRef(value)) == -1) {
+        return NULL;
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c]: This can leak `newtuple` if `PyTuple_CheckExact(obj)` is true.

File: Modules/_testlimitedcapi/tuple.c [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    NULLABLE(value);
+    if (PyTuple_CheckExact(obj)) {
+        Py_ssize_t size = PyTuple_Size(obj);
+        newtuple = PyTuple_New(size);
+        if (!newtuple) {
+            return NULL;
+        }
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SetItem(newtuple, n, Py_XNewRef(PyTuple_GetItem(obj, n)));
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c]: It is weird that we use `PyTuple_SetItem` to prepare a data for testing `PyTuple_SetItem`. And the returned value of is not checked here. But there is no other way to create a new copy of the tuple using only the limited C API.

File: Modules/_testlimitedcapi/tuple.c [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
+    }
+    if (PyTuple_SetItem(newtuple, i, Py_XNewRef(value)) == -1) {
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c]: The old value of the item is leaked here.

File: Modules/_testcapi/tuple.c [hunk 12]
```diff
@@ -2,14 +2,92 @@
...
+    }
+    NULLABLE(value);
+    if (PyTuple_CheckExact(obj)) {
+        Py_ssize_t size = PyTuple_Size(obj);
+        newtuple = PyTuple_New(size);
+        if (!newtuple) {
+            return NULL;
+        }
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SetItem(newtuple, n, Py_XNewRef(PyTuple_GetItem(obj, n)));
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/tuple.c]: It is better to use `PyTuple_SET_ITEM` and `PyTuple_GET_ITEM` here.

File: Modules/_testcapi/tuple.c [hunk 13]
```diff
@@ -2,14 +2,92 @@
...
+    Py_ssize_t newsize;
+    int new = 1;
+    if (!PyArg_ParseTuple(args, "nO|p", &newsize, &tup, &new)) {
+        return NULL;
+    }
+    if (new) {
+        Py_ssize_t size = PyTuple_Size(tup);
+        PyObject *newtup = PyTuple_New(size);
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SetItem(newtup, n, Py_XNewRef(PyTuple_GetItem(tup, n)));
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/tuple.c]: Same as above. You can also factor out the common code for creating a new copy of the tuple.

File: Modules/_testcapi/tuple.c [hunk 14]
```diff
@@ -2,14 +2,92 @@
...
+        }
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SetItem(newtuple, n, Py_XNewRef(PyTuple_GetItem(obj, n)));
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
+    }
+    PyTuple_SET_ITEM(newtuple, i, Py_XNewRef(value));
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/tuple.c]: The old value of the item is leaked.

File: Modules/_testcapi/tuple.c [hunk 15]
```diff
@@ -2,14 +2,92 @@
...
+    return newtuple;
+}
+
+static PyObject *
+tuple_resize(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    PyObject *tup;
+    Py_ssize_t newsize;
+    int new = 1;
+    if (!PyArg_ParseTuple(args, "nO|p", &newsize, &tup, &new)) {
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/tuple.c]: The order of arguments is different than in `_PyTuple_Resize()`.

File: Modules/_testcapi/tuple.c [hunk 16]
```diff
@@ -2,14 +2,92 @@
...
+            PyTuple_SetItem(newtup, n, Py_XNewRef(PyTuple_GetItem(tup, n)));
+        }
+        tup = newtup;
+    }
+    else {
+        Py_XINCREF(tup);
+    }
+    int r = _PyTuple_Resize(&tup, newsize);
+    if (r == -1) {
+        return NULL;
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/tuple.c:90]: Add an assert that `tup` is NULL here.

File: Lib/test/test_capi/test_tuple.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+        self.assertRaises(SystemError, getslice, 42, 0, 1)
+
+        # CRASHES getslice(NULL, 0, 0)
+
+    def test_tuple_setitem(self):
+        # Test PyTuple_SetItem()
+        setitem = _testlimitedcapi.tuple_setitem
+
+        tup = (0, 0)
+        self.assertEqual(setitem(tup, 0, 1), (1, 0))
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]: You need to use non-immortal values to test that the reference counting is correct, for example:
```
        tup = ([0], [0])
        self.assertEqual(setitem(tup, 0, [1]), ([1], [0]))
```
And I think that this is relevant for all or almost all other tests. They also should use non-immortal values.

Please add also tests with NULL. NULL should be accepted as a new item value, but it may be not trivial to check the result.

Comment (skirpichev) [reply Lib/test/test_capi/test_tuple.py]: @serhiy-storchaka, I hope all your review was addressed.

> but it may be not trivial to check the result.

Seems to be trivial with ctypes.  Or did I miss something? again :(

File: Modules/_testlimitedcapi/tuple.c [hunk 18]
```diff
@@ -89,7 +89,10 @@ tuple_setitem(PyObject *Py_UNUSED(module), PyObject *args)
             return NULL;
         }
         for (Py_ssize_t n = 0; n < size; n++) {
-            PyTuple_SetItem(newtuple, n, Py_XNewRef(PyTuple_GetItem(obj, n)));
+            if (PyTuple_SetItem(newtuple, n,
+                                Py_XNewRef(PyTuple_GetItem(obj, n))) == -1) {
+                return NULL;
```
Comment (serhiy-storchaka) [inline Modules/_testlimitedcapi/tuple.c:95]: Decref newtuple.

File: Lib/test/test_capi/test_tuple.py [hunk 19]
```diff
@@ -188,11 +199,19 @@ def test_tuple_setitem(self):
...
+        tup = (0, [0])
+        self.assertEqual(set_item(tup, 0, [1]), ([1], [0]))
+        self.assertEqual(set_item(tup, 1, [1]), (0, [1]))
+        tup = set_item(tup, 1, NULL)
+        func = getattr(pythonapi, "PyTuple_GetItem")
+        func.argtypes = (py_object, c_ssize_t)
+        self.assertEqual(func(tup, 1), 0)
 
     def test_tuple_resize(self):
         # Test PyTuple_Resize()
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]:
```suggestion
        # Test _PyTuple_Resize()
```

File: Lib/test/test_capi/test_tuple.py [hunk 20]
```diff
@@ -149,33 +152,41 @@ def test_tuple_getslice(self):
...
-        self.assertRaises(SystemError, getslice, [1, 2, 3], 0, 1)
+        self.assertRaises(SystemError, getslice, [1, [2], 3], 0, 1)
         self.assertRaises(SystemError, getslice, 42, 0, 1)
 
         # CRASHES getslice(NULL, 0, 0)
 
     def test_tuple_setitem(self):
         # Test PyTuple_SetItem()
+
+        import_helper.import_module('ctypes')
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py]: Many C API tests could be written with using ctypes, but ctypes is considered optional. We want to run these tests even if ctypes is not available.

You can make the wrappers of `PyTuple_GET_ITEM()` and `PyTuple_GetItem()` converting NULL to None if there is no exception. Or add a special helper function which checks whether the tuple's item is NULL. This can also be used in the test for `_PyTuple_Resize()`.

At least the code that uses ctypes can be moved to separate test methods, so some tests are still performed even without ctypes. But I think that this is more complex way.

File: Lib/test/test_capi/test_tuple.py [hunk 21]
```diff
@@ -0,0 +1,248 @@
...
+        a = ()
+        b = resize(a, 0)
+        self.assertEqual(b, a)
+        b = resize(a, 2)
+        self.assertEqual(len(b), 2)
+
+        a = (1, [2], 3)
+        b = resize(a, 3)
+        self.assertEqual(b, a)
+        b = resize(a, 2)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:240]: What happens with the reference to the last item? Make the last item a list too to test this.

File: Lib/test/test_capi/test_tuple.py [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+        self.assertEqual(b, a[:2])
+        b = resize(a, 5)
+        self.assertEqual(len(b), 5)
+        self.assertEqual(b[:3], a)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:244]: You could also check that the new items are set to NULL.

File: Lib/test/test_capi/test_tuple.py [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+        tuple_new = _testlimitedcapi.tuple_new
+        size = _testlimitedcapi.tuple_size
+
+        tup1 = tuple_new(0)
+        self.assertEqual(tup1, ())
+        self.assertEqual(size(tup1), 0)
+        self.assertIs(type(tup1), tuple)
+        tup2 = tuple_new(1)
+        self.assertIs(type(tup2), tuple)
+        self.assertEqual(size(tup2), 1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_tuple.py:58]: You can check that the new tuple is initialized with NULLs.

File: Lib/test/test_capi/test_tuple.py (hunk 20 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_tuple.py]: > Many C API tests could be written with using ctypes

Some actually are.  I did grep before trying this solution.  Shouldn't we fix that?

> but ctypes is considered optional

Hmm, not sure where it's documented.

Anyway, it worth be mentioned in https://devguide.python.org/developer-workflow/c-api/index.html#c-api-tests

> Or add a special helper function which checks whether the tuple's item is NULL.

That sounds reasonable.

File: Modules/_testlimitedcapi/tuple.c (hunk 18 above)
Comment (skirpichev) [reply Modules/_testlimitedcapi/tuple.c:95]: I will, but I doubt this branch actually is accessible.

**Comment by vstinner** [comment]
Oh, I missed this PR before I wrote my own PR: PR gh-122833.

@skirpichev: Do you plan to update your PR to fix the merge conflict?

My PR uses different code to test PyTuple_New(), PyTuple_SetItem(), PyTuple_SET_ITEM(), PyTuple_Pack() and _PyTuple_Resize(). I prefer your code for test_tuple_pack(), it tests more cases.

**Comment by skirpichev** [comment]
> Do you plan to update your PR to fix the merge conflict?

@vstinner, it seems Serhiy already did this.  Thanks.

File: Lib/test/test_capi/test_tuple.py [hunk 24]
```diff
@@ -0,0 +1,242 @@
...
+        getslice = _testlimitedcapi.tuple_getslice
+
+        # empty
+        tup = (1, 2, 3)
+        self.assertEqual(getslice(tup, PY_SSIZE_T_MIN, 0), ())
+        self.assertEqual(getslice(tup, -1, 0), ())
+        self.assertEqual(getslice(tup, 3, PY_SSIZE_T_MAX), ())
+        self.assertEqual(getslice(tup, 0, 0), ())
+        self.assertEqual(getslice(tup, 3, 0), ())
+        tup = TupleSubclass((1, 2, 3))
```
Comment (vstinner) [inline Lib/test/test_capi/test_tuple.py]: You can add check_tuple_getslice(), all it once with a tuple, and then call it with the subclass. Instead of copying/pasting the code.

File: Modules/_testcapi/tuple.c [hunk 25]
```diff
@@ -2,14 +2,107 @@
...
+    int r = _PyTuple_Resize(&tup, newsize);
+    if (r == -1) {
+        assert(tup == NULL);
+        return NULL;
+    }
+    return tup;
+}
+
+static PyObject *
+_check_item_is_NULL(PyObject *Py_UNUSED(module), PyObject *args)
```
Comment (vstinner) [inline Modules/_testcapi/tuple.c]: since the first argument is a tuple, I suggest to rename the function to: _tuple_check_item_is_NULL().

File: Lib/test/test_capi/test_tuple.py [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+    def test__tuple_resize(self):
+        # Test _PyTuple_Resize()
+        resize = _testcapi._tuple_resize
+        checknull = _testcapi._check_item_is_NULL
+
+        a = ()
+        b = resize(a, 0)
+        self.assertEqual(b, a)
+        b = resize(a, 2)
+        self.assertEqual(len(b), 2)
```
Comment (vstinner) [inline Lib/test/test_capi/test_tuple.py:233]: Can you check that items are NULL?

File: Lib/test/test_capi/test_tuple.py [hunk 27]
```diff
@@ -0,0 +1,242 @@
...
+        b = resize(a, 5)
+        self.assertEqual(len(b), 5)
+        self.assertEqual(b[:3], a)
+        self.assertTrue(all(checknull(b, i) for i in range(3, 5)))
+
+        a = ()
+        self.assertRaises(MemoryError, resize, a, PY_SSIZE_T_MAX)
+        self.assertRaises(SystemError, resize, a, -1)
+        self.assertRaises(SystemError, resize, a, PY_SSIZE_T_MIN)
+        # refcount > 1
```
Comment (vstinner) [inline Lib/test/test_capi/test_tuple.py:252]: i don't get how you make sure that the refcount is > 1 ? Maybe add self.assertGreater(sys.getrefcount(a), 1)?

File: Modules/_testcapi/tuple.c [hunk 28]
```diff
@@ -2,14 +2,107 @@
...
+        if (!newtuple) {
+            return NULL;
+        }
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SET_ITEM(newtuple, n, Py_XNewRef(PyTuple_GET_ITEM(obj, n)));
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
```
Comment (vstinner) [inline Modules/_testcapi/tuple.c]: I would prefer to copy the tuple in both cases.

File: Modules/_testcapi/tuple.c [hunk 29]
```diff
@@ -2,14 +2,107 @@
...
+    NULLABLE(value);
+    if (PyTuple_CheckExact(obj)) {
+        Py_ssize_t size = PyTuple_Size(obj);
+        newtuple = PyTuple_New(size);
+        if (!newtuple) {
+            return NULL;
+        }
+        for (Py_ssize_t n = 0; n < size; n++) {
+            PyTuple_SET_ITEM(newtuple, n, Py_XNewRef(PyTuple_GET_ITEM(obj, n)));
+        }
```
Comment (vstinner) [inline Modules/_testcapi/tuple.c]: You can move this code in a "tuple_copy()" function and reuse it in _tuple_resize().

File: Modules/_testcapi/tuple.c [hunk 30]
```diff
@@ -2,14 +2,107 @@
...
+tuple_set_item(PyObject *Py_UNUSED(module), PyObject *args)
+{
+    PyObject *obj, *value, *newtuple;
+    Py_ssize_t i;
+    if (!PyArg_ParseTuple(args, "OnO", &obj, &i, &value)) {
+        return NULL;
+    }
+    NULLABLE(value);
+    if (PyTuple_CheckExact(obj)) {
+        Py_ssize_t size = PyTuple_Size(obj);
```
Comment (vstinner) [inline Modules/_testcapi/tuple.c]:
```suggestion
        Py_ssize_t size = PyTuple_GET_SIZE(obj);
```

File: Modules/_testcapi/tuple.c (hunk 28 above)
Comment (vstinner) [inline Modules/_testcapi/tuple.c]: Don't you have to INCREF obj here?

File: Modules/_testcapi/tuple.c [hunk 31]
```diff
@@ -2,14 +2,107 @@
...
+            PyTuple_SET_ITEM(newtuple, n, Py_XNewRef(PyTuple_GET_ITEM(obj, n)));
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
+    }
+    PyObject *val = PyTuple_GET_ITEM(newtuple, i);
+    PyTuple_SET_ITEM(newtuple, i, Py_XNewRef(value));
+    Py_DECREF(val);
```
Comment (vstinner) [inline Modules/_testcapi/tuple.c]:
```suggestion
    Py_XDECREF(val);
```

File: Modules/_testlimitedcapi/tuple.c [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+            if (PyTuple_SetItem(newtuple, n,
+                                Py_XNewRef(PyTuple_GetItem(obj, n))) == -1) {
+                Py_XDECREF(newtuple);
+                return NULL;
+            }
+        }
+    }
+    else {
+        NULLABLE(obj);
+        newtuple = obj;
```
Comment (vstinner) [inline Modules/_testlimitedcapi/tuple.c]: Don't you have to INCREF obj here?

**Review by serhiy-storchaka — APPROVED** [review]
I went through this code very carefully and fixed several errors related to refcounting. Also added some new checks and tweaked the style a bit.

LGTM.

File: Lib/test/test_capi/test_tuple.py (hunk 27 above)
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_tuple.py:252]: One reference from the local variable and other from a function argument.

Actually, it is > 1 in all other tests too.

**Review by vstinner — APPROVED** [review]
LGTM

File: Lib/test/test_capi/test_tuple.py (hunk 24 above)
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_tuple.py]: To many hassle, in my opinion. For subclass we only need few basic tests, just to be sure that they are accepted.

File: Modules/_testcapi/tuple.c (hunk 25 above)
Comment (serhiy-storchaka) [reply Modules/_testcapi/tuple.c]: Done.

File: Lib/test/test_capi/test_tuple.py (hunk 26 above)
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_tuple.py:233]: Done.

File: Modules/_testcapi/tuple.c (hunk 28 above)
Comment (serhiy-storchaka) [reply Modules/_testcapi/tuple.c]: How do you copy an instance of a tuple subclass? Or non-tuple? We need to test this API with tuple subclasses and non-tuples.

Comment (serhiy-storchaka) [reply Modules/_testcapi/tuple.c]: Not here, but before returning result. I fixed this in other way.

File: Modules/_testcapi/tuple.c (hunk 31 above)
Comment (serhiy-storchaka) [reply Modules/_testcapi/tuple.c]: It is not NULL.

File: Modules/_testlimitedcapi/tuple.c (hunk 32 above)
Comment (serhiy-storchaka) [reply Modules/_testlimitedcapi/tuple.c]: Same as in other place.

File: Lib/test/test_capi/test_tuple.py (hunk 27 above)
Comment (vstinner) [reply Lib/test/test_capi/test_tuple.py:252]: Ah ok, it makes sense.

**Comment by vstinner** [comment]
I merged your PR, thanks @skirpichev!

**Comment by skirpichev** [comment]
@vstinner, shouldn't this be backported to 3.12 and 3.13?

**Comment by vstinner** [comment]
> @vstinner, shouldn't this be backported to 3.12 and 3.13?

I would prefer to not backport them.

@serhiy-storchaka: What's your call on backporting new C API tests to stable branches?

**Comment by serhiy-storchaka** [comment]
I prefer to backport tests as far as possible. This helps if we change the behavior in new version or backport bug fixes.

**Comment by skirpichev** [comment]
I'm working on backports.

**Comment by skirpichev** [comment]
3.12 is less trivial due to changes from #116417.

I can do backport for current state of 3.12, but maybe it does make sense to backport first pr's from above issue?  Or 3.12 not worth efforts at all?
